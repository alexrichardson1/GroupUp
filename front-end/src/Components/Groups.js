import { Button, Col, Container, Row } from "react-bootstrap";
import { LinkContainer } from "react-router-bootstrap";
import "bootstrap/dist/css/bootstrap.min.css";
import { useParams } from "react-router-dom";
import { useEffect, useState } from "react";
import Filter from "./Filter";
import Group from "./Group";
import axios from "axios";
// import { Link } from "react-router-dom";
import { useState, useEffect } from "react";
import Filter from "./Filter";
import hackathons from "../data/projects.json";

async function getGroups() {
  var result = [];
  await axios
    .get("http://localhost:5000/group")
    .then((res) => {
      const groups = res.data;
      result = groups;
    })
    .catch((error) => {
      console.log(error);
    });
  return result;
}

const Groups = () => {
  const { id } = useParams();
  const hackathonId = parseInt(id);

  const [allGroups, setAllGroups] = useState([]);

  const setGroups = async () => {
    const nonFilteredGroups = await getGroups();
    const filteredGroups = nonFilteredGroups.filter(
      (group) => group.projectid === hackathonId
    );
    setAllGroups(filteredGroups);
    setFilteredGroups(filteredGroups);
  };

  const hackathonReqs = hackathons.filter((proj) => proj.id === hackathonId)[0]
    .requirements;

  const [filteredGroups, setFilteredGroups] = useState([]);
  const [activeFilters, setActiveFilters] = useState(new Map());

  useEffect(() => {
    setGroups();
  });

  useEffect(() => {
    for (const [k, v] of activeFilters.entries()) {
      setFilteredGroups((f) =>
        f.filter((group) => group.requirements[k] === v)
      );
    }

    console.log("Active Filters after: ");

    for (const [k, v] of activeFilters.entries()) {
      console.log(k, v);
    }
    console.log("--------------------");
  }, [activeFilters]);

  const updateFilter = (key, value) => {
    setActiveFilters((prev) => new Map(prev).set(key, value));
  };

  const removeFilter = (key) => {
    setActiveFilters((prev) => {
      const newState = new Map(prev);
      newState.delete(key);
      return newState;
    });
  };

  const filterGroupsOnReq = (reqName, reqVal) => {
    setFilteredGroups(allGroups);
    console.log("--------------------");
    console.log(
      "attempting filter of {" + reqName + "} and with value {" + reqVal + "}"
    );
    console.log("Active Filters before: ");
    updateFilter(reqName, reqVal);
  };

  const getAllReqVars = (req) => {
    const vars = new Set();
    hackathonGroups.forEach((group) => vars.add(group.requirements[req]));
    return [...vars];
  };

  const deleteFilter = (reqName) => {
    // activeFilters.delete(reqName);
    removeFilter(reqName);
    const newFilteredGroups = allGroups;
    for (const [k, v] of activeFilters.entries()) {
      newFilteredGroups.filter((group) => group.requirements[k] === v);
    }
    setFilteredGroups(newFilteredGroups);
  };

  return (
    <div>
      <Container>
        <Row>
          <h3>{filteredGroups.length} Groups looking for members:</h3>
        </Row>
        <Row>
          <LinkContainer to="/createGroup">
            <Button>Advertise my group!</Button>
          </LinkContainer>
        </Row>
        <Row>
          <Col>
            <h1>{activeFilters.size}</h1>
            {hackathonReqs.map((req) => (
              <Filter
                requirementName={req}
                requirementsList={getAllReqVars(req)}
                filterFunction={filterGroupsOnReq}
                resetFunction={deleteFilter}
              />
            ))}
          </Col>
          <Col>
            {filteredGroups.map((group) => (
              <Group group={group} key={group.id} />
            ))}
          </Col>
        </Row>
      </Container>
    </div>
  );
};

export default Groups;
