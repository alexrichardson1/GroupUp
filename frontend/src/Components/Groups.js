import Group from "Components/Group";
import { Button, Col, Container, Row } from "react-bootstrap";
import { LinkContainer } from "react-router-bootstrap";
import "bootstrap/dist/css/bootstrap.min.css";
import { useParams } from "react-router-dom";
import { useState, useEffect } from "react";
import Filter from "Components/Filter";
import hackathons from "data/projects.json";
import Navbar from "Components/NavBar";

const Groups = ({ allGroups }) => {
  const { id } = useParams();
  const hackathonId = parseInt(id);
  const hackathonName = hackathons.filter((proj) => proj.id === hackathonId)[0]
    .name;
  const hackathonGroups = allGroups.filter(
    (group) => group.projectId === hackathonId
  );
  const hackathonReqs = hackathons.filter((proj) => proj.id === hackathonId)[0]
    .requirements;

  const [filteredGroups, setFilteredGroups] = useState([...hackathonGroups]);
  const [activeFilters, setActiveFilters] = useState(new Map());

  useEffect(() => {
    for (const [k, v] of activeFilters.entries()) {
      setFilteredGroups((f) =>
        f.filter((group) => group.requirements[k] === v)
      );
    }

    console.log("Active Filters after: ");

    for (const [k, v] of activeFilters.entries()) {
      console.log(k, v);
    }
    console.log("--------------------");
  }, [activeFilters]);

  const updateFilter = (key, value) => {
    setActiveFilters((prev) => new Map(prev).set(key, value));
  };

  const removeFilter = (key) => {
    setActiveFilters((prev) => {
      const newState = new Map(prev);
      newState.delete(key);
      return newState;
    });
  };

  /* This will filter groups on all "active filters", plus the new filter specified. */
  const filterGroupsOnReq = (reqName, reqVal) => {
    setFilteredGroups(hackathonGroups);
    console.log("--------------------");
    console.log(
      "attempting filter of {" + reqName + "} and with value {" + reqVal + "}"
    );
    console.log("Active Filters before: ");
    updateFilter(reqName, reqVal);
  };

  const getAllReqVars = (req) => {
    const vars = new Set();
    hackathonGroups.forEach((group) => vars.add(group.requirements[req]));
    return [...vars];
  };

  const deleteFilter = (reqName) => {
    removeFilter(reqName);
    const newFilteredGroups = hackathonGroups;
    for (const [k, v] of activeFilters.entries()) {
      newFilteredGroups.filter((group) => group.requirements[k] === v);
    }
    setFilteredGroups(newFilteredGroups);
  };

  return (
    <div>
      <Navbar renderBool={[true, true, true, false]} create={false} />
      <Container>
        <Row>
          <h3>
            {filteredGroups.length} Groups looking for members in:â€„
            {hackathonName}
          </h3>
        </Row>
        <Row>
          <LinkContainer to="/createGroup">
            <Button>Advertise my group!</Button>
          </LinkContainer>
        </Row>
        <Row>
          <Col>
            {hackathonReqs.map((req) => (
              <Filter
                requirementName={req}
                requirementsList={getAllReqVars(req)}
                filterFunction={filterGroupsOnReq}
                resetFunction={deleteFilter}
              />
            ))}
          </Col>
          <Col>
            {filteredGroups.map((group) => (
              <Group group={group} key={group.id} />
            ))}
          </Col>
        </Row>
      </Container>
    </div>
  );
};

export default Groups;
