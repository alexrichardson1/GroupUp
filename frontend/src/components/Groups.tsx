import { Component } from "react";
import Group from "components/Group";
import { Button, Col, Container, Modal, Row } from "react-bootstrap";
import { LinkContainer } from "react-router-bootstrap";
import "bootstrap/dist/css/bootstrap.min.css";
import Navbar from "components/NavBar";
import Filter from "components/Filter";
import { UserContext } from "components/auth/UserContext";
import { GroupT, ProjectT } from "types/types";
import { dummyProject, getGroups, getProject, setFilters } from "common/api";

interface Props {
  id: number;
}

interface State {
  groups: GroupT[];
  projects: ProjectT;
  requirements: string[];
  filteredGroups: GroupT[];
  activeFilters: Map<string, string>;
  modalShow: boolean;
}

export default class Groups extends Component<Props, State> {
  constructor(props: Props) {
    super(props);
    this.state = {
      groups: [],
      projects: dummyProject,
      requirements: [],
      filteredGroups: [],
      activeFilters: new Map(),
      modalShow: false,
    };
  }

  handleClose() {
    this.setState({ modalShow: false });
  }
  handleShow() {
    this.setState({ modalShow: true });
    for (const [k, v] of this.state.activeFilters.entries()) {
      console.log(k + " -> " + v);
    }
  }

  groupIsFull(group: GroupT) {
    return 1 + group.teammates.length >= group.maxmembers;
  }

  override async componentDidMount() {
    this.setState({ groups: await getGroups() });
    this.setState({ projects: await getProject(this.props.id) });
    this.setState({ filteredGroups: await getGroups() });
    this.setState({ requirements: this.hackathonReqs() });
    document.title = "Available Listings";
  }

  hackathonReqs = () => {
    return this.state.projects.requirements;
  };

  allReqVars = (reqIndex: number) => {
    const vars: string[] = [];
    this.state.groups.forEach((g) => {
      const requirement = g.requirements[reqIndex];
      if (requirement) {
        vars.push(requirement);
      }
    });
    return [...new Set(vars)];
  };

  filterGroupsOnReq = (reqName: string, reqVar: string) => {
    this.addFilter(reqName, reqVar);
    this.setState(
      (state) => ({ filteredGroups: state.groups }),
      () => this.refilter()
    );
  };

  refilter = () => {
    for (const [, v] of this.state.activeFilters.entries()) {
      let newFilteredGroups = [];
      if (v === "Any") {
        newFilteredGroups = [...this.state.filteredGroups];
      } else {
        for (let i = 0; i < this.state.filteredGroups.length; i++) {
          if (
            this.state.filteredGroups[i] &&
            //@ts-ignore
            this.state.filteredGroups[i].requirements.includes(v)
          ) {
            newFilteredGroups.push(this.state.filteredGroups[i]);
          }
        }
      }
      //@ts-ignore
      this.state.filteredGroups = newFilteredGroups;
      this.setState({ ...this.state });
    }
  };

  addFilter = (key: string, value: string) => {
    this.setState({ activeFilters: this.state.activeFilters.set(key, value) });
  };

  removeFilter = (key: string) => {
    this.filterGroupsOnReq(key, "Any");
  };

  createFilterComponents = () => {
    const comps: JSX.Element[] = [];
    this.state.requirements.forEach((req) => {
      comps.push(
        <Filter
          requirementName={req}
          requirementsList={this.allReqVars(
            this.state.requirements.indexOf(req)
          )}
          filterFunction={this.filterGroupsOnReq}
          resetFunction={this.removeFilter}
        />
      );
    });
    return comps;
  };

  saveFilters = (email: string) => {
    console.log(email);
  };

  postEmail = (email: string) => {
    const vals = [];
    for (const [, v] of this.state.activeFilters.entries()) {
      vals.push(v);
    }
    setFilters(email, vals);
  };

  override render() {
    return (
      <UserContext.Consumer>
        {({ email }) => (
          <div>
            <Navbar renderBool={[true, true, true, false]} create={false} />
            <Modal show={this.state.modalShow} onHide={this.handleClose}>
              <Modal.Header>
                <Modal.Title>Saved</Modal.Title>
              </Modal.Header>

              <Modal.Body>
                <p>
                  Your current filtered search has been saved. Find it under
                  your account.
                </p>
              </Modal.Body>

              <Modal.Footer>
                <Button variant="secondary" onClick={() => this.handleClose()}>
                  Close
                </Button>
              </Modal.Footer>
            </Modal>
            <Container>
              <Row>
                <Col>
                  <h3>
                    {this.state.filteredGroups.length} Groups looking for
                    members in:â€„
                    {this.state.projects.name}
                  </h3>
                </Col>
                <Col className="advertCol align-items-center">
                  <Button
                    className="advertBtn"
                    onClick={() => {
                      this.handleShow();
                      this.postEmail(email);
                    }}
                  >
                    Save this search
                  </Button>
                  <LinkContainer to={`/createGroup/${this.props.id}`}>
                    <Button className="advertBtn">Advertise my group</Button>
                  </LinkContainer>
                </Col>
              </Row>
              {/* <Row> */}

              {/* </Row> */}
              <Row>
                <Col className="filterCol">
                  <h5>Filters:</h5>
                  {this.createFilterComponents()}
                </Col>
                <Col className="listingsCol">
                  {this.state.filteredGroups.map((group) => (
                    <Group
                      group={group}
                      requirementNames={this.state.requirements}
                      projectId={this.props.id}
                      key={group.id}
                    />
                  ))}
                </Col>
              </Row>
            </Container>
          </div>
        )}
      </UserContext.Consumer>
    );
  }
}
