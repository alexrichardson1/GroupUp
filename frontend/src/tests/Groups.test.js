import renderer from "react-test-renderer";
import { BrowserRouter } from "react-router-dom";
import Groups from "components/Groups";

test("renders correctly", () => {
  const tree = renderer
    .create(
      <BrowserRouter>
        <Groups />
      </BrowserRouter>
    )
    .toJSON();
  expect(tree).toMatchSnapshot();
});
