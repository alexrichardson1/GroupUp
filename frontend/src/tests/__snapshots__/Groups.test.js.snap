// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<div>
  <nav
    className="navBar navbar navbar-expand-lg navbar-light fixed-top"
  >
    <span
      className="navbar-brand"
      style={
        Object {
          "color": "white",
        }
      }
    >
      <img
        alt="group up logo"
        className="logoImage"
        href="/home"
        onClick={[Function]}
        src="https://i.imgur.com/w92t1xH.png"
      />
    </span>
    <button
      aria-controls="basic-navbar-nav"
      aria-label="Toggle navigation"
      className="navbar-toggler collapsed"
      onClick={[Function]}
      type="button"
    >
      <span
        className="navbar-toggler-icon"
      />
    </button>
    <div
      aria-expanded={null}
      className="navbar-collapse collapse"
    >
      <div
        className="mr=auto navbar-nav"
        onKeyDown={[Function]}
      >
        <span
          className="navbar-brand"
          style={
            Object {
              "color": "rgb(238, 237, 237)",
            }
          }
        >
           
        </span>
        <a
          className="nav-link"
          data-rb-event-key="/home"
          href="/home"
          onClick={[Function]}
          onKeyDown={[Function]}
          style={
            Object {
              "color": "rgb(238, 237, 237)",
            }
          }
        >
          Home
          <img
            alt="drop down arrow"
            className="processImage"
            src="/right-caret.svg"
          />
        </a>
        <a
          className="nav-link"
          data-rb-event-key="/selection"
          href="/selection"
          onClick={[Function]}
          onKeyDown={[Function]}
          style={
            Object {
              "color": "rgb(238, 237, 237)",
            }
          }
        >
          Select Hackathon
          <img
            alt="drop down arrow"
            className="processImage"
            src="/right-caret.svg"
          />
        </a>
        <span
          className="navbar-brand"
          style={
            Object {
              "color": "rgb(238, 237, 237)",
            }
          }
        >
          Group Listings
        </span>
      </div>
      <div
        className="ms-auto navbar-nav"
        onKeyDown={[Function]}
      >
        <a
          className="ml-auto navHelp navbar-brand"
          href="/help"
          id="collasible-nav-dropdown"
          onClick={[Function]}
          title="Help"
        >
          <img
            alt="help"
            className="splitterImage"
            src="/info-circle.svg"
          />
        </a>
      </div>
    </div>
  </nav>
  <div
    className="container"
  >
    <div
      className="row"
    >
      <div
        className="col"
      >
        <h3>
          0
           Groups looking for members in:â€„
          IC HACK X
        </h3>
      </div>
      <div
        className="advertCol align-items-center col"
      >
        <button
          className="advertBtn btn btn-primary"
          disabled={false}
          onClick={[Function]}
          type="button"
        >
          Save this search
        </button>
        <a
          className="advertBtn btn btn-primary"
          href="/createGroup/undefined"
          onClick={[Function]}
          onKeyDown={[Function]}
        >
          Advertise my group
        </a>
      </div>
    </div>
    <div
      className="row"
    >
      <div
        className="filterCol col"
      >
        <h5>
          Filters:
        </h5>
      </div>
      <div
        className="listingsCol col"
      />
    </div>
  </div>
</div>
`;
